storage: Change sequence cache to an abort cache

spencerkimball opened the pull request:
… txn

Nothing is dependent on the potentially many sequence cache entries
per transaction so this is just a simplification and for certain use
cases may yield a performance improvement.

Added a unittest for copying sequence caches when they may now have
an overlapping entry which must be parsed to determine which entry
has the highest epoch and sequence numbers. The previous code had a
bug with overcounting stats in the event of overlap.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5626)

<!-- Reviewable:end -->

spencerkimball committed:
Simplify the sequence cache code by removing the multiple entries per txn

Nothing is dependent on the potentially many sequence cache entries
per transaction so this is just a simplification and for certain use
cases may yield a performance improvement.

Added a unittest for copying sequence caches when they may now have
an overlapping entry which must be parsed to determine which entry
has the highest epoch and sequence numbers. The previous code had a
bug with overcounting stats in the event of overlap.

tschottdorf commented:
I'm not convinced about this change - sure, the simplification here using `SequenceCacheEntry` instead of the individual fields decoded from the key is nice, but that just has to decrease performance since now every sequence cache check must decode the whole thing. Additionally, you've added a lot of reads for copying the sequence cache, which sits in the critical path for a split, and more work done for updating the sequence cache entries which is also in the critical path for every write. Maybe none of it matters? I don't know, but with the concerns about sequence cache performance we have I'm uncomfortable not backing this up with data first.

We can get all of the benefits of deduplicating sequence cache entries if we do it as a side-effect on `ResolveIntent`. There's no need to push this in this deep as far as I can tell.

---

Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/replica.go, line 1628 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KDx3ytevIBgVapMIZ3U:-KDx3ytfuB-YRFi3jkV2:-1158024737) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b50ba5c7ddef1443d263608971dd88b974ea833/storage/replica.go#L1628)):*
I indicated this in an earlier PR, but there's usually no need to unmarshal the entry before these changes - shouldn't we keep it that way, especially if we're chasing performance?

---

*[storage/sequence_cache.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KDx4GR1DJxjK7hcXq5w:-KDx4GR2qBf7YieI1B0p:930031677) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b50ba5c7ddef1443d263608971dd88b974ea833/storage/sequence_cache.go#L42)):*
Can't we give back an empty entry instead?

---

*[storage/sequence_cache.go, line 174 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KDx59oG8GP5onEH9xLc:-KDx59oG8GP5onEH9xLd:1618839117) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b50ba5c7ddef1443d263608971dd88b974ea833/storage/sequence_cache.go#L174)):*
This seems like it would do way more work in the critical split path - before we were simply copy-pasting (as much as possible), and now we added a read for each of the keys. I'm not sure that this is a net improvement.

---

*[storage/sequence_cache.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KDx5T0ODUI3clr1Y6VV:-KDx5T0ODUI3clr1Y6VW:1475449712) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b50ba5c7ddef1443d263608971dd88b974ea833/storage/sequence_cache.go#L182)):*
Can we reuse any of the `updateStatsOnSomething` helpers for this and the (old) stats update below?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-:-KDx5wRShCQovjFJU2Ex:-2093427002)*
<!-- Sent from Reviewable.io -->

spencerkimball committed:
Change sequence cache to an abort cache

Instead of always writing to the sequence cache on successive writes
we now only keep track of aborted transactions. This simplifies
some of the code paths as well.

It works by moving the transaction sequence number into the `TxnMeta`,
which is stored with intents. Now replays instead of failing at the
start of a raft command by checking against the sequence cache will
instead fail by trying to write over a pre-existing intent from the
same transaction but with an earlier or equal sequence number.

But there's a wrinkle, which is how to deal with multiple commands
within a single batch, which all have the same sequence number
(there's just one sequence number per batch in the batch header),
but whose keys might overlap. An example would be two increments
to the same key. We deal with this by storing an additional int32
in the `TxnMeta`, called `BatchIndex`, which is incremented on
each successive command in the batch.

This change creates virtually no garbage for GC.

Fixes #5088

spencerkimball commented:
Fixes #5639

spencerkimball commented:
@tschottdorf your previous comments are all outdated by the most recent commit.

spencerkimball commented on storage/abort_cache.go:
```diff
@@ -0,0 +1,199 @@
+// Copyright 2014 The Cockroach Authors.
```
This file and `abort_cache_test` are mostly just code movement with the various bits of cleanup to pare out the unnecessary fields from the original `SequenceCacheEntry`.

spencerkimball commented:
It's actually a surprisingly small change...mostly renaming here.

spencerkimball commented:
@maxlang any idea what's up with the build failure for circleci?

cuongdo commented:
A published version of an NPM module got yanked with no notice. Rebase from
master and the problem should go away.

On Tue, Mar 29, 2016 at 2:57 AM Spencer Kimball <notifications@github.com>
wrote:

> @maxlang <https://github.com/maxlang> any idea what's up with the build
> failure for circleci?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5626#issuecomment-202744707>
>

tschottdorf commented:
Partial review, will continue after lunch.

---

Reviewed 32 of 32 files at r3, 21 of 32 files at r4.
Review status: 21 of 32 files reviewed at latest revision, 18 unresolved discussions.

---

*[keys/constants.go, line 88 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2Cd8XyJylVWJ5Ei1X:-KE2Cd8XyJylVWJ5Ei1Y:-1698740489) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/keys/constants.go#L88)):*
Too late to rename the `res-` to `ac-`?

---

*[roachpb/data.proto, line 312 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2DwHA94Qs7OKOVToF:-KE2DwHA94Qs7OKOVToG:729130985) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/roachpb/data.proto#L312)):*
s/transactions/a transaction/

---

*[roachpb/data.proto, line 313 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2Dxk0nUsmhh1ondmX:-KE2Dxk0nUsmhh1ondmY:-1252660872) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/roachpb/data.proto#L313)):*
s/have/has/, s/They're/It is/

---

*[roachpb/data.proto, line 314 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2E3IL8I03NphdQqgM:-KE2E3IL8I03NphdQqgN:-672045961) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/roachpb/data.proto#L314)):*
s/it.*with intents/the range may have contained intents of the aborted txn/

---

*[roachpb/data.proto, line 316 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2EIoAM0RfZzDumC7J:-KE2EIoAM0RfZzDumC7K:-354490752) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/roachpb/data.proto#L316)):*
try something like

contains.*\n.*\.n.* informs the transaction that it has aborted and must start fresh with an updated priority.

---

*[sql/testdata/poison_after_push, line 60 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2EjH1U4BRUD0O8JVk:-KE2EjH1U4BRUD0O8JVl:-742477060) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/sql/testdata/poison_after_push#L60)):*
Undo that accidental deletion.

---

*[storage/abort_cache.go, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2Eq7lQM4KBlVNSe5q:-KE2Eq7lQM4KBlVNSe5r:-2070257862) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L34)):*
s/protection .*/protection against an aborted but active transaction not reading values it wrote (due to its intents having been removed)./

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2FFxSLcw01JQgsyXr:-1767743074) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
Can we use updateXonY?

---

*[storage/abort_cache.go, line 184 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FMlBLKwE8FIxprs-:-KE2FMlBLKwE8FIxprs0:1772408594) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L184)):*
s/abort number/entry/

---

*[storage/gc_queue.go, line 233 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FrRiLHbKXeQ5kKhF:-KE2FrRiLHbKXeQ5kKhG:-295079570) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/gc_queue.go#L233)):*
> // TODO(tschottdorf): update to strategy in #5639.

---

*[storage/intent_resolver.go, line 270 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2GdD6gB2epyoLrQF9:-KE2GdD6gB2epyoLrQFA:1918183223) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/intent_resolver.go#L270)):*
it's ok to remove the TODO, but then this really ought to be `true` (after all, some intents might be getting aborted, but the client may not know). Add a comment explaining that, too.

---

*[storage/intent_resolver.go, line 287 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2H0xusHOQL6sfqv5j:-KE2H0xusHOQL6sfqv5k:-516327721) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/intent_resolver.go#L287)):*
s/poison/!poison/ and add a comment why it's ok to clear out the seq cache.

---

*[storage/intent_resolver.go, line 294 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2H9bGmRpH4Zbl0wbQ:-KE2H9bGmRpH4Zbl0wbR:464803544) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d2f39daa993f233a14df1467234734da9a6ffc1/storage/intent_resolver.go#L294)):*
Is this not happening any more?

---

*[storage/replica.go, line 1603 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2Hayic7YuhQVJ-608:-KE2Hayic7YuhQVJ-609:1355038701) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/replica.go#L1603)):*
Make sure that you've removed that error which was returned if `!aborted` earlier. Maybe you did, but I think I saw it in some late-ish diff.

---

*[storage/replica_command.go, line 610 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2C4A8MjhliiKKAapU:-KE2C4A9DKgTU9ymBaYe:-497754427) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/replica_command.go#L610)):*
GC'ed

---

*[storage/replica_command.go, line 1182 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2AwONxeZcVzJrujET:-KE2AwOOhlSeLQ7Rn_ap:-1911416060) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/replica_command.go#L1182)):*
Poisoning should also be conditional on `intent.Status == roachpb.ABORTED`, right? `{set,clear}AbortCache` could fuse into `setAbortCache(batch, ms, args.IntentTxn, args.Poison)` where the last parameter indicates whether to poison or to clear a poisoned entry.

```go
if intent.Status == roachpb.ABORTED {
  return reply, r.setAbortCache(batch, ms, args.IntentTxn, args.Poison)
}
```

---

*[storage/replica_command.go, line 1207 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2Bay3OCgno5BM2jh9:-KE2Bay4fQggh2sUzIig:-1677837463) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/replica_command.go#L1207)):*
ditto.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-:-KE2I0xxsujAl-1Efg8P:536267715)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 21 of 32 files reviewed at latest revision, 18 unresolved discussions.

---

*[keys/constants.go, line 88 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2Cd8XyJylVWJ5Ei1X:-KE2ZvmzovKWU0jFry0o:-609445857) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/keys/constants.go#L88)):*
Changed.

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2aE9INU21yxInmM65:354328557) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
What's that?

---

*[storage/intent_resolver.go, line 294 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2H9bGmRpH4Zbl0wbQ:-KE2czozB_rq5zjl87Zk:670032284) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d2f39daa993f233a14df1467234734da9a6ffc1/storage/intent_resolver.go#L294)):*
It is still happening. I will restore this comment.

---

*[storage/replica.go, line 1603 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2Hayic7YuhQVJ-608:-KE2dKYRuvA57sl0H6T7:-1705903436) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/replica.go#L1603)):*
You mean the `errNoCacheEntry` error? It's a goner.

---

*[storage/replica_command.go, line 610 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2C4A8MjhliiKKAapU:-KE2dUoKmWYxZyRhJw9j:-129971194) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/replica_command.go#L610)):*
Renamed as GC'd.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 11 of 32 files at r4, 8 of 8 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[keys/constants.go, line 86 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2g28FolUlvcuvlitY:-KE2g28G1IfiHwzYxlhu:1748617595) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f029e23cbdc453c694747bcaba0a026617f601a/keys/constants.go#L86)):*
s/used to protect/used for the mechanism which protects/

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2gEglUeV-bmVavMdy:57782731) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
I mean the various helpers in mvcc.go which usually perform those stats updates, to get rid of ad-hoc computations here. Not urgent - add a TODO (maybe for @nvanbenschoten).

---

*[storage/intent_resolver.go, line 270 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2GdD6gB2epyoLrQF9:-KE2gwOkDnvXQV9wtvEr:-1799871999) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/intent_resolver.go#L270)):*
Here's an example comment:

> In this code path, we're resolving intents outside of the context of an EndTransaction. Naively, it doesn't seem like we need to poison the abort cache since we're pushing with PUSH_TOUCH - meaning that the primary way our Push can only lead to aborting intents is that of the transaction having timed out (and thus presumably no client being around any more, though at the time of writing we don't guarantee that). But there's another path in which the Push comes back successful, namely that of the transaction already having been aborted by someone else, in which case the client may still be running. Thus, we must poison.

---

*[storage/intent_resolver.go, line 287 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2H0xusHOQL6sfqv5j:-KE2iUXRk7YxspNjjf5p:-40811827) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/intent_resolver.go#L287)):*
Suggested addition to comment:

> This mechanism assumes that when an EndTransaction fails, the client makes no assumptions about the true result. For example, an attempt to explicitly rollback the transaction may succeed (triggering this code path), but the result may not make it back to the client.

---

*[storage/intent_resolver.go, line 294 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2H9bGmRpH4Zbl0wbQ:-KE2gZs5WYk7vSyGz-_w:104500654) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d2f39daa993f233a14df1467234734da9a6ffc1/storage/intent_resolver.go#L294)):*
Maybe it shouldn't happen. After all, what is it going to clean out? Nothing on commit, nothing on explicit abort.

---

*[storage/replica_command.go, line 612 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2gkKfQ1IKyBT74zYM:-KE2gkKfQ1IKyBT74zYN:1107856751) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f029e23cbdc453c694747bcaba0a026617f601a/storage/replica_command.go#L612)):*
inline comment on `false`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Basically <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> when the remaining (minor) comments are addressed.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/intent_resolver.go, line 270 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2GdD6gB2epyoLrQF9:-KE2jp9UcCiZNM4he07f:-321189258) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/intent_resolver.go#L270)):*
some awkward wording in my comment above, might wanna give this another pass through your brain.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-:-KE2kNYMI4LPsZp51ZsU:-1631357979)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2nb3NfY0BuW0WMBXs:609674920) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
I didn't see anything appropriate in there. They all deal with more complicated compositions involving updates and none of them are exported...

---

*[storage/intent_resolver.go, line 270 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2GdD6gB2epyoLrQF9:-KE2mhSw0hI_Qk3TL6eY:-1839607857) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/intent_resolver.go#L270)):*
Good, this comports pretty exactly with the comment I was just writing (but yours is more descriptive), so we're on the same page.

---

*[storage/intent_resolver.go, line 294 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2H9bGmRpH4Zbl0wbQ:-KE2nDnU150BMwOzN_pj:501538592) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d2f39daa993f233a14df1467234734da9a6ffc1/storage/intent_resolver.go#L294)):*
Oh, good point. Removed comment.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2nxtbB1tNBX2IkE33:-186765875) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
Well, can be changed - but for now let's stick with the TODO.

---

*[storage/replica.go, line 1644 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2ndFSl4WUpgl6vB2u:-KE2ndFT-Tkm_5lSWzWD:-2041943874) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f029e23cbdc453c694747bcaba0a026617f601a/storage/replica.go#L1644)):*
Oh, didn't see this before. This is kind of a nasty way to plumb this. Let me take a quick look if there's a way to get something now and not have an annoying migration after.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 28 of 32 files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2p3vMeGi1jGUr60KY:-730568633) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
Well this code already breaks an abstraction by fiddling directly with mvcc stuff in the interests of efficiency. I think the suggested TODO to further break the abstraction might make things worse. How about a comment saying something similar?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 28 of 32 files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2pVjOcK6_mdJQBVAq:-1335364699) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
Yeah, sounds acceptable to me. Now that this cache is much smaller, maybe this efficiency thing can be backed out. That could be a comment (but still with a TODO).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 28 of 32 files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2q3Af73GClk6brDHm:1976954926) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
Actually, I added a TODO after some thinking. Why aren't we just using `MVCCIterate` and `MVCCPut` here? Would be way simpler and would avoid breaking the abstraction. I just don't want to change as part of this PR.

---

*[storage/replica.go, line 1644 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2ndFSl4WUpgl6vB2u:-KE2qGl-f9h57cpGAbd1:-490122963) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f029e23cbdc453c694747bcaba0a026617f601a/storage/replica.go#L1644)):*
OK...

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/abort_cache.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2FFxRA-06UiJS6oJd:-KE2rN53BCa59b-wD2oJ:1455284288) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bed253f8fcd2b9872e98e7140e0e47f678ab8c8/storage/abort_cache.go#L150)):*
Yeah, this is what I suggested in my last comment for the TODO.

---

*[storage/intent_resolver.go, line 295 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2rwROhsyJWYKLNkxT:-KE2rwROhsyJWYKLNkxU:-1627018078) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f029e23cbdc453c694747bcaba0a026617f601a/storage/intent_resolver.go#L295)):*
I don't see a corresponding edit in `EndTransaction` - should back out the call to `setAbortCache` as well.

---

*[storage/intent_resolver.go, line 275 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2rjG0RE3x60PGUFPj:-KE2rjG1APLrVLblNbD9:-1556389108) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7cf107f93711dac0f0a6ad6bd9359d942f080bb/storage/intent_resolver.go#L275)):*
> the primary way in which our Push leads to aborting intents

---

*[storage/replica.go, line 1644 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2ndFSl4WUpgl6vB2u:-KE2rTSJGh6bO5OIO4Yg:-300572402) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f029e23cbdc453c694747bcaba0a026617f601a/storage/replica.go#L1644)):*
Sorry, this isn't actually bad - we store that stuff in the intent, so that's perfectly fine. All we need to figure out is the copy-on-write story, but nothing there has changed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 23 of 32 files at r4, 5 of 8 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[keys/constants.go, line 86 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2g28FolUlvcuvlitY:-KE2pSzJ3Xy22iLrzjVL:409250258) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f029e23cbdc453c694747bcaba0a026617f601a/keys/constants.go#L86)):*
The grammar here is bugging me.

---

*[keys/keys.go, line 105 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2q4ouslh7ic9EGi94:-KE2q4ouslh7ic9EGi95:1689102443) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7cf107f93711dac0f0a6ad6bd9359d942f080bb/keys/keys.go#L105)):*
this comment doesn't make sense to me. what detail is specified?

---

*[sql/testdata/poison_after_push, line 1 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2slTb9cb3K_z1QE-y:-KE2slTb9cb3K_z1QE-z:-583605658) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7cf107f93711dac0f0a6ad6bd9359d942f080bb/sql/testdata/poison_after_push#L1)):*
you lost the word transaction for some reason

---

*[storage/intent_resolver.go, line 275 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2r0oZbDtGRCBQ8ZE0:-KE2r0oa06dmQZw59neL:-87181455) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7cf107f93711dac0f0a6ad6bd9359d942f080bb/storage/intent_resolver.go#L275)):*
what does the word "only" mean here? as opposed to having additional side effects? what additional side effects?

---

*[storage/intent_resolver.go, line 302 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2rGShVhYguyK_2Asx:-KE2rGShVhYguyK_2Asy:1537764625) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7cf107f93711dac0f0a6ad6bd9359d942f080bb/storage/intent_resolver.go#L302)):*
s/true result/result/

---

*[storage/replica_test.go, line 1582 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2tnUoEeOyIkPOS3JT:-KE2tnUpfDU3aaUi1sxj:-891363896) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7cf107f93711dac0f0a6ad6bd9359d942f080bb/storage/replica_test.go#L1582)):*
why not just check the error?

---

*[storage/store.go, line 386 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2u0BlhJDM-MHuvyxG:-KE2u0BlhJDM-MHuvyxH:1835465659) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d2f39daa993f233a14df1467234734da9a6ffc1/storage/store.go#L386)):*
is this still used anywhere? I saw some place that was setting it was removed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[keys/keys.go, line 105 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626#-KE2q4ouslh7ic9EGi94:-KE3EWO1pya6_sdR3VZi:1715104529) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7cf107f93711dac0f0a6ad6bd9359d942f080bb/keys/keys.go#L105)):*
heh, that's pretty unclear.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5626)*
<!-- Sent from Reviewable.io -->
